base/intfuncs.jl

```diff
@@ -602,74 +602,111 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
 ## integer to string functions ##
 
 function bin(x::Unsigned, pad::Integer, neg::Bool)
-    i = neg + max(pad,sizeof(x)<<3-leading_zeros(x))
-    a = StringVector(i)
+    m = 8sizeof(x) - leading_zeros(x)::Int
+    n = neg + max((pad % Int)::Int, m)
```

Comment:
Using `pad % Int` is invalid in some corner cases (e.g. for `pad == typemax(Int)+1`). It's not necessarily a problem for `bin` as it's an internal function, but some form of validation should probably remain somewhere, e.g. in client functions like `string` (which could take care of converting to `Int`).

Reply:
Is that a security issue?

Reply:
I don't know, but this makes the behavior of `string` sometimes incorrect, in an  and observable way.

Reply:
> an observable way

Can you give a specific example? Being secure is a requirement of the correctness, but otherwise I don't think the correct behavior is obvious.

Reply:
Currently, `string(0, pad=typemin(Int128)+3)` returns `""`, which seems correct as per the docstring. With this change (if I'm not mistaken), it will return `"000"`, which seems wrong.

Reply:
I will try to fix it. However, this PR starts with avoiding invalidation in the first place.

Reply:
> However, this PR starts with avoiding invalidation in the first place.

I don't understand how this is related. I'm not familiar with invalidations hunting, but hopefully we can both avoid invalidations and be correct. In this case, what about `pad = clamp(pad, Int)::Int` ?

Reply:
It's "usually" not a bad idea. However, I don't really want to limit it to the range of `Int` (since it is too "wide" for practical use cases). I want to just avoid the invalidation. 
The `pad`s are always `Int`, in the use cases in `Base` i.e. in the default sysimage. Therefore, `pad % Int` is just insurance. In that respect, I think `clamp` is not better than `rem`.

Perhaps a reasonable fix would be to get rid of the `% Int` (i.e. reverting). Do you have any thoughts?

Reply:
> I don't really want to limit it to the range of Int (since it is too "wide" for practical use cases)

I don't really understand this sentence. In my view, everything which is `<0` can be set to `0`, and everything which is bigger than `typemax(Int)` probably makes no sense has it's not practical to have a string of length `> typemax(Int)`, so such number can be replaced by `typemax(Int)`. I just described `clamp`. What doesn't seem OK is to use `pad % Int` as this has a weird behavior of for these extreme values of `pad`.

> The pads are always Int, in the use cases in Base i.e. in the default sysimage. 

Maybe, but I'm concerned by usage outside of base where your current change leads to an incorrect result as I showed with an example. But unfortunately I have no other suggestion, as I didn't follow how invalidation work, and I don't know which is better, `clamp` or the `max` that was there or writing the equivalent of `clamp` by hand (which would be sad) etc.

Reply:
>> I don't really want to limit it to the range of Int (since it is too "wide" for practical use cases)
>
> I don't really understand this sentence.

I meant a user friendly error message or docstring.